text/0000-less-unwinding.md

```diff
@@ -0,0 +1,281 @@
+* Embedded Rust libraries can avoid taking down the entire process and can
+  instead gracefully inform the caller that an internal logic error occurred.
+* Rust applications can isolate failure from various components. The classical
+  example of this is Servo can display a "red X" for an image which fails to
+  decode instead of aborting the entire browser or killing an entire page.
+
+While these are examples where a recoverable panic is useful, there are many
+applications where recovering panics is undesirable or doesn't lead to anything
+productive:
+
+* Rust applications which use `Result` for error handling typically use `panic!`
+  to indicate a fatal error, in which case the process *should* be taken down.
+* Many applications simply can't recover from an internal assertion failure, so
+  there's no need trying to recover it.
+* To implement a recoverable panic, the compiler and standard library use a
+  method called stack unwinding. The compiler must generate code to support this
+  unwinding, however, and this takes time in codegen and optimizers.
+* Low-level applications typically don't use unwinding at all as there's no
+  stack unwinder (e.g. kernels).
+
+> **Note**: as an idea of the compile-time and object-size savings from
+> disabling the extra codegen, compiling Cargo as a library is 11% faster (16s
+> from 18s) and 13% smaller (15MB to 13MB). Sizable gains!
+
+Overall, the ability to recover panics is something that needs to be decided at
+the application level rather than at the language level. Currently the compiler
+does not support the ability to translate panics to process aborts in a stable
+fashion, and the purpose of this RFC is to add such a venue.
+
+With such an important codegen option, however, as whether or not exceptions can
+be caught, it's easy to get into a situation where libraries of mixed
+compilation modes are linked together, causing odd or unknown errors. This RFC
+proposes a situation similar to the design of custom allocators to alleviate
+this situation.
+
+# Detailed design
+[design]: #detailed-design
+
+The major goal of this RFC is to develop a work flow around managing crates
+which wish to disable unwinding. This intends to set forth a complete vision for
+how these crates interact with the ecosystem at large. Much of this design will
+be similar to the [custom allocator RFC][custom-allocators].
+
+[custom-allocators]: https://github.com/rust-lang/rfcs/blob/master/text/1183-swap-out-jemalloc.md
+
+### High level design
+
+This section serves as a high-level tour through the design proposed in this
+RFC. The linked sections provide more complete explanation as to what each step
+entails.
+
+* The compiler will have a [new stable flag](#new-compiler-flags), `-C panic`
+  which will configure how unwinding-related code is generated.
+* [Two new unstable attributes](#panic-attributes) will be added to the
+  compiler, `#![needs_panic_runtime]` and `#![panic_runtime]`. The standard
```

Comment:
@nagisa can you elaborate on what you're thinking? Given the way these attributes are defined below I'm not understanding how this could be implemented with `#[cfg]`.